File: src/markdown.ts
/**
 * Convert markdown to Telegram-compatible HTML
 * Supports: bold, italic, code, code blocks, links
 */

// Escape HTML special characters
function escapeHtml(text: string): string {
	return text
		.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;");
}

// Convert markdown to Telegram HTML
export function markdownToHtml(text: string): string {
	// Use null character as delimiter to avoid matching bold/italic patterns
	// The previous __CODE_BLOCK_0__ format was incorrectly matched by the
	// bold pattern /__([^_]+)__/g, causing placeholders to be converted to
	// <b>CODE_BLOCK_0</b> instead of being preserved for later restoration.
	const PLACEHOLDER_START = "\x00";
	const PLACEHOLDER_END = "\x00";

	// First, extract and preserve code blocks to prevent processing inside them
	const codeBlocks: string[] = [];
	let processed = text.replace(
		/```(\w*)\n?([\s\S]*?)```/g,
		(_, _lang, code) => {
			const index = codeBlocks.length;
			// Escape HTML inside code blocks
			codeBlocks.push(`<pre>${escapeHtml(code.trim())}</pre>`);
			return `${PLACEHOLDER_START}CODE_BLOCK_${index}${PLACEHOLDER_END}`;
		},
	);

	// Extract inline code
	const inlineCodes: string[] = [];
	processed = processed.replace(/`([^`]+)`/g, (_, code) => {
		const index = inlineCodes.length;
		inlineCodes.push(`<code>${escapeHtml(code)}</code>`);
		return `${PLACEHOLDER_START}INLINE_CODE_${index}${PLACEHOLDER_END}`;
	});

	// Escape HTML in remaining text
	processed = escapeHtml(processed);

	// Convert markdown formatting
	// Bold: **text** or __text__
	processed = processed.replace(/\*\*([^*]+)\*\*/g, "<b>$1</b>");
	processed = processed.replace(/__([^_]+)__/g, "<b>$1</b>");

	// Italic: *text* or _text_ (but not inside words)
	processed = processed.replace(
		/(?<![a-zA-Z])\*([^*]+)\*(?![a-zA-Z])/g,
		"<i>$1</i>",
	);
	processed = processed.replace(
		/(?<![a-zA-Z])_([^_]+)_(?![a-zA-Z])/g,
		"<i>$1</i>",
	);

	// Strikethrough: ~~text~~
	processed = processed.replace(/~~([^~]+)~~/g, "<s>$1</s>");

	// Links: [text](url)
	processed = processed.replace(
		/\[([^\]]+)\]\(([^)]+)\)/g,
		'<a href="$2">$1</a>',
	);

	// Restore code blocks
	for (let i = 0; i < codeBlocks.length; i++) {
		processed = processed.replace(
			`${PLACEHOLDER_START}CODE_BLOCK_${i}${PLACEHOLDER_END}`,
			codeBlocks[i],
		);
	}

	// Restore inline code
	for (let i = 0; i < inlineCodes.length; i++) {
		processed = processed.replace(
			`${PLACEHOLDER_START}INLINE_CODE_${i}${PLACEHOLDER_END}`,
			inlineCodes[i],
		);
	}

	return processed;
}

// Strip all markdown formatting to plain text
export function stripMarkdown(text: string): string {
	return (
		text
			// Remove code blocks
			.replace(/```[\s\S]*?```/g, (match) => {
				const code = match.replace(/```\w*\n?/, "").replace(/```$/, "");
				return code.trim();
			})
			// Remove inline code backticks
			.replace(/`([^`]+)`/g, "$1")
			// Remove bold
			.replace(/\*\*([^*]+)\*\*/g, "$1")
			.replace(/__([^_]+)__/g, "$1")
			// Remove italic
			.replace(/\*([^*]+)\*/g, "$1")
			.replace(/_([^_]+)_/g, "$1")
			// Remove strikethrough
			.replace(/~~([^~]+)~~/g, "$1")
			// Convert links to just text
			.replace(/\[([^\]]+)\]\([^)]+\)/g, "$1")
	);
}

File: src/markdown.rs
fn escape_html(text: &str) -> String {
    text.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

pub fn markdown_to_html(text: &str) -> String {
    // Use null character as delimiter for placeholders (same approach as TS version)
    const PH: char = '\x00';

    // Extract and preserve code blocks
    let mut code_blocks: Vec<String> = Vec::new();
    let mut processed = {
        let re = regex::Regex::new(r"```(\w*)\n?([\s\S]*?)```").unwrap();
        re.replace_all(text, |caps: &regex::Captures| {
            let idx = code_blocks.len();
            let code = caps.get(2).map_or("", |m| m.as_str()).trim();
            code_blocks.push(format!("<pre>{}</pre>", escape_html(code)));
            format!("{PH}CODE_BLOCK_{idx}{PH}")
        })
        .into_owned()
    };

    // Extract inline code
    let mut inline_codes: Vec<String> = Vec::new();
    processed = {
        let re = regex::Regex::new(r"`([^`]+)`").unwrap();
        re.replace_all(&processed, |caps: &regex::Captures| {
            let idx = inline_codes.len();
            let code = caps.get(1).map_or("", |m| m.as_str());
            inline_codes.push(format!("<code>{}</code>", escape_html(code)));
            format!("{PH}INLINE_CODE_{idx}{PH}")
        })
        .into_owned()
    };

    // Escape HTML in remaining text
    processed = escape_html(&processed);

    // Bold: **text** or __text__
    let re = regex::Regex::new(r"\*\*([^*]+)\*\*").unwrap();
    processed = re.replace_all(&processed, "<b>$1</b>").into_owned();
    let re = regex::Regex::new(r"__([^_]+)__").unwrap();
    processed = re.replace_all(&processed, "<b>$1</b>").into_owned();

    // Italic: *text* or _text_ (bold ** already processed above)
    let re = regex::Regex::new(r"\*([^*]+)\*").unwrap();
    processed = re.replace_all(&processed, "<i>$1</i>").into_owned();
    let re = regex::Regex::new(r"\b_([^_]+)_\b").unwrap();
    processed = re.replace_all(&processed, "<i>$1</i>").into_owned();

    // Strikethrough: ~~text~~
    let re = regex::Regex::new(r"~~([^~]+)~~").unwrap();
    processed = re.replace_all(&processed, "<s>$1</s>").into_owned();

    // Links: [text](url)
    let re = regex::Regex::new(r"\[([^\]]+)\]\(([^)]+)\)").unwrap();
    processed = re
        .replace_all(&processed, r#"<a href="$2">$1</a>"#)
        .into_owned();

    // Restore code blocks
    for (i, block) in code_blocks.iter().enumerate() {
        processed = processed.replace(&format!("{PH}CODE_BLOCK_{i}{PH}"), block);
    }

    // Restore inline code
    for (i, code) in inline_codes.iter().enumerate() {
        processed = processed.replace(&format!("{PH}INLINE_CODE_{i}{PH}"), code);
    }

    processed
}

pub fn strip_markdown(text: &str) -> String {
    let mut result = text.to_string();

    // Remove code blocks (keep content)
    let re = regex::Regex::new(r"```\w*\n?([\s\S]*?)```").unwrap();
    result = re.replace_all(&result, "$1").into_owned();

    // Remove inline code backticks
    let re = regex::Regex::new(r"`([^`]+)`").unwrap();
    result = re.replace_all(&result, "$1").into_owned();

    // Remove bold
    let re = regex::Regex::new(r"\*\*([^*]+)\*\*").unwrap();
    result = re.replace_all(&result, "$1").into_owned();
    let re = regex::Regex::new(r"__([^_]+)__").unwrap();
    result = re.replace_all(&result, "$1").into_owned();

    // Remove italic
    let re = regex::Regex::new(r"\*([^*]+)\*").unwrap();
    result = re.replace_all(&result, "$1").into_owned();
    let re = regex::Regex::new(r"_([^_]+)_").unwrap();
    result = re.replace_all(&result, "$1").into_owned();

    // Remove strikethrough
    let re = regex::Regex::new(r"~~([^~]+)~~").unwrap();
    result = re.replace_all(&result, "$1").into_owned();

    // Convert links to just text
    let re = regex::Regex::new(r"\[([^\]]+)\]\([^)]+\)").unwrap();
    result = re.replace_all(&result, "$1").into_owned();

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bold() {
        assert_eq!(markdown_to_html("**bold**"), "<b>bold</b>");
        assert_eq!(markdown_to_html("__bold__"), "<b>bold</b>");
    }

    #[test]
    fn test_italic() {
        assert_eq!(markdown_to_html("*italic*"), "<i>italic</i>");
        assert_eq!(markdown_to_html("_italic_"), "<i>italic</i>");
    }

    #[test]
    fn test_code_block() {
        assert_eq!(
            markdown_to_html("```rust\nfn main() {}\n```"),
            "<pre>fn main() {}</pre>"
        );
    }

    #[test]
    fn test_inline_code() {
        assert_eq!(markdown_to_html("`code`"), "<code>code</code>");
    }

    #[test]
    fn test_html_escaping() {
        assert_eq!(markdown_to_html("<script>"), "&lt;script&gt;");
    }

    #[test]
    fn test_html_in_code_block() {
        assert_eq!(
            markdown_to_html("```\n<div>test</div>\n```"),
            "<pre>&lt;div&gt;test&lt;/div&gt;</pre>"
        );
    }

    #[test]
    fn test_link() {
        assert_eq!(
            markdown_to_html("[click](https://example.com)"),
            r#"<a href="https://example.com">click</a>"#
        );
    }

    #[test]
    fn test_strikethrough() {
        assert_eq!(markdown_to_html("~~deleted~~"), "<s>deleted</s>");
    }

    #[test]
    fn test_code_blocks_not_processed() {
        let input = "```\n**not bold** _not italic_\n```";
        let output = markdown_to_html(input);
        assert!(output.contains("**not bold**"));
        assert!(!output.contains("<b>"));
    }

    #[test]
    fn test_strip_markdown_bold() {
        assert_eq!(strip_markdown("**bold**"), "bold");
    }

    #[test]
    fn test_strip_markdown_link() {
        assert_eq!(strip_markdown("[text](url)"), "text");
    }

    #[test]
    fn test_strip_markdown_code_block() {
        assert_eq!(strip_markdown("```\ncode\n```"), "code\n");
    }

    #[test]
    fn test_plain_text_unchanged() {
        assert_eq!(markdown_to_html("hello world"), "hello world");
    }
}
